### PR TITLE
Support EIP-1193 compliant providers

### LINKED ISSUE #58: WalletConnect: auto-track wallet CONNECT, DISCONNECT and CHAIN_CHANGED events
Add an additional optional parameter to `init()` that takes a provider to watch blockchain events. We should be able to support the Metamask and WalletConnect providers.

```javascript
ArcxAnalyticsSdk.init(
  apiKey: string, 
  config?: Partial<SdkConfig>, 
  initialProvider?: providers.Web3Provider
)
```

It looks like we can listen for `chainChanged`, `accountsChanged` and `disconnect` events on the provider itself with this method: https://docs.walletconnect.com/1.0/quick-start/dapps/web3-provider#events-eip-1193

If we do this, we also have to add a `setProvider(provider: providers.Web3Provider)` function.

### LINKED ISSUE #59: WalletConnect: auto-track transactions

### LINKED ISSUE #60: WalletConnect: auto-track signatures

### PR DESCRIPTION
**Pull request by gtupak**
Closes #58
Closes #59
Closes #60

This PR accepts an additional `initialProvider` in the config, as well as adding a `setProvider(provider: EIP1193)` function that allows tracking on any provider that is [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) compliant.

### CODE REVIEW

File: example/cra-provider/package.json [hunk 1]
```diff
@@ -20,7 +23,7 @@
     "web-vitals": "^2.1.0"
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "GENERATE_SOURCEMAP=false react-scripts start",
```
Comment (gtupak) [inline example/cra-provider/package.json:26]: Without this, a bunch of warnings are thrown at server start because of the wallet-connect library

File: example/cra-provider/src/MetamaskButtons.tsx [hunk 2]
```diff
@@ -1,54 +0,0 @@
-import { useWeb3React } from '@web3-react/core'
```
Comment (gtupak) [inline example/cra-provider/src/MetamaskButtons.tsx:1]: This file effectively became `EthereumEventsButtons.tsx`

File: src/ArcxAnalyticsSdk.ts [hunk 3]
```diff
@@ -41,32 +53,30 @@ export class ArcxAnalyticsSdk {
...
-
-    if (this.sdkConfig.trackTransactions) {
-      this._trackTransactions()
-    }
-
-    if (this.sdkConfig.trackSigning) {
-      this._trackSigning()
-    }
+    this._handleAccountDisconnected = this._handleAccountDisconnected.bind(this)
+    this.setProvider = this.setProvider.bind(this)
```
Comment (gtupak) [inline src/ArcxAnalyticsSdk.ts:57]: These functions are used inside listeners and use `this.`, which is lost during the event call. For this reason, I'm binding `this` to them.

File: src/ArcxAnalyticsSdk.ts [hunk 4]
```diff
@@ -148,9 +158,14 @@ export class ArcxAnalyticsSdk {
...
-        'ArcxAnalyticsSdk::_handleAccountDisconnected: previousChainId or previousConnectedAccount is not set',
-      )
+      /**
+       * It is possible that this function has already been called once and the cached values
+       * have been cleared. This can happen in the follwoing scenario:
+       * 1. Initialize ArcxAnalyticsProvider with the default config (sets MM as the initial provider)
+       * 2. Connect WalletConnect
+       * 3. Disconnect
+       */
+      return
```
Comment (gtupak) [inline src/ArcxAnalyticsSdk.ts:168]: Weird edge-case. For some reason, the disconnect event is being emitted 3 times when disconnecting wallet-connect and I'm not sure why. I spend a considerable time on it and decided to move on. This is not the case with metamask.

File: src/ArcxAnalyticsSdk.ts [hunk 5]
```diff
@@ -201,12 +216,18 @@ export class ArcxAnalyticsSdk {
...
     }
 
-    const request = provider.request
+    if (!this._originalRequest) {
+      this._originalRequest = provider.request
+    }
+
+    // Deliberately not using this._original request to not intefere with the signature tracking's
+    // request modification
+    const request = provider.request.bind(provider)
```
Comment (gtupak) [inline src/ArcxAnalyticsSdk.ts:230]: Without `.bind(provider)`, the request call fails as it loses its link to the provider

File: src/ArcxAnalyticsSdk.ts [hunk 6]
```diff
@@ -227,12 +248,18 @@ export class ArcxAnalyticsSdk {
...
-    const request = provider.request
-    provider.request = async ({ method, params }: RequestArguments) => {
+
+    if (!this._originalRequest) {
+      this._originalRequest = this.provider.request
+    }
+
+    // Deliberately not using this._original request to not intefere with the transaction tracking's
+    // request modification
+    const request = this.provider.request.bind(this.provider)
```
Comment (gtupak) [inline src/ArcxAnalyticsSdk.ts:261]: Same as above

**Review by gtupak — COMMENTED** [review]
Added some explanations to make the review easier. Sorry for the big line changes on the tests, quite a few edge-cases

File: example/cra-provider/src/Web3Buttons.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      sdk.setProvider(provider.provider as any)
+    } else if (sdk && sdk.provider && !provider) {
+      // provider disconnected from external source (e.g. user clicked disconnect in wallet connect)
+      setConnectedWallet(undefined)
+    }
+  }, [sdk, provider])
+
+  const onConnectWalletClicked = async (givenConnector: Connector, chainId?: number) => {
+    try {
+      console.log({ givenConnector: givenConnector, provider: givenConnector.provider })
```
Comment (xitronix) [inline example/cra-provider/src/Web3Buttons.tsx:23]: Did you leave this comment with purpose?

File: example/cra-provider/src/connectors/walletConnect.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import { initializeConnector } from '@web3-react/core'
+import { WalletConnect } from '@web3-react/walletconnect'
+
+export const [walletConnect, walletConnectHooks] = initializeConnector<WalletConnect>((actions) => {
+  const mainnetKey = process.env.REACT_APP_ALCHEMY_KEY_MAINNET
```
Comment (xitronix) [inline example/cra-provider/src/connectors/walletConnect.ts:5]: why don't just use getEnv from @arcxmoney/core

File: src/ArcxAnalyticsSdk.ts (hunk 6 above)
Comment (xitronix) [reply src/ArcxAnalyticsSdk.ts:261]: how do we lose link to request if provider is binded and has proper reference?

File: src/ArcxAnalyticsSdk.ts [hunk 9]
```diff
@@ -253,10 +280,62 @@ export class ArcxAnalyticsSdk {
     return true
   }
 
+  private _initializeWeb3Tracking() {
+    if (this.provider) {
```
Comment (xitronix) [inline src/ArcxAnalyticsSdk.ts:284]: Invert if statement to throw error or send report if provider does not exist

**Review by xitronix — CHANGES_REQUESTED** [review]
Rebase and i'll review again

File: example/cra-provider/src/Web3Buttons.tsx (hunk 7 above)
Comment (gtupak) [reply example/cra-provider/src/Web3Buttons.tsx:23]: no, removed

File: example/cra-provider/src/connectors/walletConnect.ts (hunk 8 above)
Comment (gtupak) [reply example/cra-provider/src/connectors/walletConnect.ts:5]: I don't want to add a package with all these deps just to use `getEnv`: 
```
  "dependencies": {
    "@arcxgame/contracts": "^11.4.0",
    "@google-cloud/storage": "^5.16.1",
    "@restless/ethereum": "^0.3.0",
    "@restless/restless": "^0.4.2",
    "@restless/sanitizers": "^0.2.5",
    "axios": "^0.21.1",
    "dayjs": "^1.10.6",
    "dotenv": "^8.2.0",
    "ethers": "5.5.4",
    "ethers-multicall": "^0.2.3",
    "lodash": "^4.17.21",
    "mongodb": "^4.2.0",
    "node-cache": "^5.1.2",
    "rollbar": "^2.24.0"
  },
```

Comment (xitronix) [reply example/cra-provider/src/connectors/walletConnect.ts:5]: Just create one then :) With getting env throwing if doesn't exist. Useful utility

Comment (gtupak) [reply example/cra-provider/src/connectors/walletConnect.ts:5]: As discussed, won't create it bc it's just used here

**Review by xitronix — APPROVED** [review]
LGTM. Checked with @gtupak , so should work

**Review by xitronix — CHANGES_REQUESTED** [review]
Update cra-script-tag

**Comment by gtupak** [comment]
@xitronix for the script tag example to work, we first need to merge and deploy this branch. I created the requested adjustment in https://github.com/arcxmoney/analytics-sdk/pull/80

**Review by xitronix — APPROVED** [review]
LGTM
